file_formats/cmt_format.ml

```diff
@@ -60,8 +72,9 @@ type cmt_infos = {
   cmt_imports : (string * Digest.t option) list;
   cmt_interface_digest : Digest.t option;
   cmt_use_summaries : bool;
-  cmt_uid_to_loc : Location.t Shape.Uid.Tbl.t;
+  cmt_uid_to_decl : item_declaration Shape.Uid.Tbl.t;
   cmt_impl_shape : Shape.t option; (* None for mli *)
+  cmt_shape_index : (Shape.t, Longident.t Location.loc list) Hashtbl.t
```

Comment:
I'm surprised this is a table rather than just a list or array of uses. IIUC The consumers of these are going to have to iterate over all these shapes anyway since they need to reduce them with the non-local shapes to get the relavent `Uid.t` for each of them. Since they are only needed for iteration the extra structure and bookkeeping of a table seems unnecessary. Also values of type `Shape.t` can be arbitrarily large, so aren't great keys for a table.

Reply:
I think you could also use a type like:
```ocaml
type use =
  | Resolved of Uid.t option
  | Unresolved of Shape.t
```
since you are only interested on getting the relevant `Uid.t` for each use, so once you've found it you can throw away the rest of the shape. Obviously this only affects module uses -- shapes for values and types have no structure beyond the `Uid.t` anyway.

Reply:
Done in 29171e8